lib/matplotlib/testing/decorators.py

```diff
@@ -387,35 +387,45 @@ def decorator(func):
 
         _, result_dir = _image_directories(func)
 
-        if len(inspect.signature(func).parameters) == 2:
-            # Free-standing function.
-            @pytest.mark.parametrize("ext", extensions)
-            def wrapper(ext):
-                fig_test = plt.figure("test")
-                fig_ref = plt.figure("reference")
-                func(fig_test, fig_ref)
-                test_image_path = result_dir / (func.__name__ + "." + ext)
-                ref_image_path = (
-                    result_dir / (func.__name__ + "-expected." + ext))
-                fig_test.savefig(test_image_path)
-                fig_ref.savefig(ref_image_path)
-                _raise_on_image_difference(
-                    ref_image_path, test_image_path, tol=tol)
-
-        elif len(inspect.signature(func).parameters) == 3:
-            # Method.
-            @pytest.mark.parametrize("ext", extensions)
-            def wrapper(self, ext):
-                fig_test = plt.figure("test")
-                fig_ref = plt.figure("reference")
-                func(self, fig_test, fig_ref)
-                test_image_path = result_dir / (func.__name__ + "." + ext)
-                ref_image_path = (
-                    result_dir / (func.__name__ + "-expected." + ext))
-                fig_test.savefig(test_image_path)
-                fig_ref.savefig(ref_image_path)
-                _raise_on_image_difference(
-                    ref_image_path, test_image_path, tol=tol)
+        @pytest.mark.parametrize("ext", extensions)
+        def wrapper(*args, ext, **kwargs):
+            fig_test = plt.figure("test")
+            fig_ref = plt.figure("reference")
+            func(*args, fig_test=fig_test, fig_ref=fig_ref, **kwargs)
+            test_image_path = result_dir / (func.__name__ + "." + ext)
+            ref_image_path = result_dir / (
+                func.__name__ + "-expected." + ext
+            )
+            fig_test.savefig(test_image_path)
+            fig_ref.savefig(ref_image_path)
+            _raise_on_image_difference(
+                ref_image_path, test_image_path, tol=tol
+            )
+
+        sig = inspect.signature(func)
+        new_sig = sig.replace(
+            parameters=(
+                [
```

Comment:
As a side note, I find your code formatting style extremely hard to read.

Reply:
I agree this just looks like the worst of `black`, things like
```
wrapper.pytestmark = get_unpacked_marks(
    wrapper
) + get_unpacked_marks(func)
```
just completely obfuscate the structure of the code -- compare with
```
wrapper.pytestmark = (get_unpacked_marks(wrapper)
                      + get_unpacked_marks(func))
```
(or with an additional newline after the opening parenthesis).

Reply:
This is indeed black's doing.  

Ironically, the biggest selling point of black for me is avoiding exactly these conversations.

Reply:
The fact that it's automated doesn't absolve it from completely violating basic esthetics.

Reply:
fwiw @NelleV and I have been trying to come up with a better automated scheme and it's certainly not trivial... Nevertheless, since afaict matplotlib hasn't added black as a dependency, it's a bit weird to add badly black-formatted code...

Reply:
(And @tacaswell it looks like you've set it to 71c wide? Which is unconventional to say the least...)

Reply:
(I personally just don't buy into the idea that code can be properly formatted without understanding the underlying semantics.)

Reply:
(Juan is considering doing constrained reinforcement learning to learn the rules…)

Reply:
I thought I set it to 80c (to keep our pyflakes happy).   Black does a slightly better job if you give it a bit more width to work with.